stream/src/public_key/cipher.rs

```diff
@@ -0,0 +1,312 @@
+use crate::Error;
+use chacha20poly1305::{ChaCha20Poly1305, KeyInit, KeySizeUser};
+use commonware_cryptography::{CoreSha256, Hasher, Sha256};
+use hkdf::{hmac::digest::typenum::Unsigned, Hkdf};
+use zeroize::Zeroize;
+
+// The size of the key used by the ChaCha20Poly1305 cipher.
+const CHACHA_KEY_SIZE: usize = <ChaCha20Poly1305 as KeySizeUser>::KeySize::USIZE;
+
+// A constant prefix used for the salt hash in the HKDF key derivation.
+const BASE_KDF_PREFIX: &[u8] = b"commonware-stream/KDF/v1/";
+
+/// Key Derivation Function (KDF) to derive directional ChaCha20Poly1305 ciphers using HKDF-SHA256.
+///
+/// This function derives two ChaCha20Poly1305 ciphers based on:
```

Comment:
Wonder if it makes sense to be explicit that this is an "ordered" list (and that is critical for the shared secret to be generated correctly by both).